src/matlab2tikz.m

```diff
@@ -963,23 +884,107 @@ case guitypes()
+        for lhandle = m2t.legendHandles(:)'
+            ud = get(lhandle, 'UserData');
+            % Empty if no legend and multiple handles if plotyy
+            if ~isempty(ud) && any(handle == ud.handle)
+                legendhandle = lhandle;
+                break
+            end
+        end
+    case 'MATLAB'
+        legendhandle = legend(handle);
+end
+
+% NOTE: there is a BUG in HG1 and Octave. Setting the box off sets the
+% legend visibility off too. We assume the legend is visible if it has 
+% a visible child.
+isInvisibleHG2 = isHG2() && ~isVisible(legendhandle);
+isInvisibleHG1orOctave = (~isHG2() || strcmpi(env,'Octave')) &&...
+    ~isVisibleContainer(legendhandle);
+
+% Do not return the handle if legend is invisible
+if isInvisibleHG1orOctave || isInvisibleHG2;
     legendhandle = [];
-    switch getEnvironment
-        case 'Octave'
-            % Make sure that m2t.legendHandles is a row vector.
-            for lhandle = m2t.legendHandles(:)'
-                ud = get(lhandle, 'UserData');
-                if isVisibleContainer(lhandle) && any(handle == ud.handle)
-                    legendhandle = lhandle;
-                    break;
-                end
+end
+end
+% ==============================================================================
+function entries = getLegendEntries(m2t)
+entries = [];
+legendHandle = m2t.axesContainers{end}.LegendHandle;
+
+if isempty(legendHandle), return, end
+
+switch getEnvironment
+    case 'Octave'
+        % See set(hlegend, "deletefcn", {@deletelegend2, ca, [], [], t1, hplots}); in legend.m
+        delfun  = get(legendHandle,'deletefcn');
+        entries = delfun{6};
+        
+        % Bubble-up legend entry properties from child to hggroup root
+        % for guessable objects
+        guessable = {'specgraph.errorbarseries','matlab.graphics.chart.primitive.ErrorBar',...
+                     'specgraph.scattergroup'  ,'matlab.graphics.chart.primitive.Scatter'};
+        for ii = 1:numel(entries)
+            child = entries(ii);
+            anc   = ancestor(child,'hggroup');
+            if isempty(anc)
+                continue
             end
-        case 'MATLAB'
-            % no action needed
-        otherwise
-            errorUnknownEnvironment();
-    end
+            cl = guessOctavePlotType(anc);
+            if any(strcmpi(cl, guessable))
+                legendString = get(child,'displayname');
+                set(anc,'displayname',legendString);
+                entries(ii) = anc;
+            end
+        end
+        
+    case 'MATLAB'
+        % Undocumented property (exists at least since 2008a)
+        entries = double(get(legendHandle,'PlotChildren'));
+        
+        % Take only the first child from a pure hggroup (e.g. bodeplots)
+        for ii = 1:numel(entries)
+            entry     = entries(ii);
+            isHggroup = any(strcmpi(get(entry,'Type'),{'hggroup', 'matlab.graphics.primitive.Group'})) & ...
```

Comment:
Wouldn't that be confusing since we usually handle hggroups which have a special *series* class? Also, the function would not get much re-use, since in handleAllChildren we use a `switch` statement.